src/sentry/api/endpoints/artifact_lookup.py

```diff
@@ -165,6 +168,46 @@ def get(self, request: Request, project: Project) -> Response:
         return Response(serialize(found_artifacts, request.user))
 
 
+def renew_artifact_bundles(used_artifact_bundles: Mapping[int, datetime]):
+    # We take a snapshot in time that MUST be consistent across all updates.
+    now = datetime.now(tz=pytz.UTC)
+
+    for (bundle_id, date_added) in used_artifact_bundles.items():
+        if ArtifactBundle.can_be_renewed(date_added):
+            # In case this bundle can be renewed we will try to renew all the instances connected to this bundle with
+            # the same `date_added`.
+            for model in (
+                ArtifactBundle,
+                ProjectArtifactBundle,
+                DebugIdArtifactBundle,
+                ReleaseArtifactBundle,
+            ):
+                async_update_model(model=model, bundle_id=bundle_id, new_date_added=now)
+
+
+def async_update_model(model, bundle_id: int, new_date_added: datetime):
+    id_key = "id" if model == ArtifactBundle else "artifact_bundle_id"
+    # We want to asynchronously update the field. This is automatically handled in a batching fashion by the buffer.
+    #
+    # This implementation might have race conditions in which a `buffer_incr` task will be scheduled for the same
+    # `bundle_id` in different times. We assume here that there is some ordering in execution and tasks are completed
+    # in a FIFO fashion without interleaving updates. In case this will turn out to not be the case, we might have
+    # the database into an inconsistent state, where the `date_added` field of the entities connected to the same
+    # `bundle_id` are different.In addition, we don't expect the same `bundle_id` to be request in close sequence since
+    # Symbolicator caches for ~1 hour the bundles received by this endpoint.
+    #
+    # In case we will end up with inconsistent `date_added` into the database, we might have to build a fully custom
+    # batching implementation with proper ordering and consistency guarantees.
```

Comment:
I would rather solve this problem by updating more frequently, as discussed in chat.

If the database is partitioned by date, and the oldest partition is deleted regularly, we just have to make sure the bundle ends up in one of the other partitions. which ones exactly does not matter that much, as long as it is being touched *at all*.

Or maybe my understanding is wrong here, and you could end up with some `JOIN` problems across partitions?

Reply:
This problem is not solved by updating more frequently, my concern was about interleaving updates with different `date_added` but for the same entities.

Reply:
you are explicitly setting it to `new_date_added = datetime.now()`, you are not adding a delta to it.
so at most you have a couple of seconds difference but that shouldn’t matter